packages/material-ui/src/index.d.ts

```diff
@@ -64,6 +64,18 @@ export type ConsistentWith<T, U> = Pick<U, keyof T & keyof U>;
  */
 export type Overwrite<T, U> = Omit<T, keyof U> & U;
 
+/**
+ * a property P will will have the the type of:
+ * - DecorationTargetProps if it is not present in InjectedProps
+ * - InjectedTargetProps if it is present in DecorationTargetProps
+ *
+ */
+export type Matching<InjectedProps, DecorationTargetProps> = {
+  [P in keyof DecorationTargetProps]: P extends keyof InjectedProps
+    ? InjectedProps[P]
```

Comment:
This no longer validates that the two interfaces passed in are matching. If `InjectedProps[P]` doesn't extend `DecorationTargetProps[P]` (assuming their keys are shared) then that means that the two definitions are incompatible and cannot be reconciled. E.g., this would accept: `Matching<{ test: string; }, { test: boolean; }>`

Reply:
I posted a new definition for Matching in my PR that is IMO slightly more correct, at least as far as `connect` goes. Not sure if you have different requirements here as I don't use material-ui.

Reply:
@apapirovski 
I agree that approach is indeed better and I tried this before but this caused some regression. I realized that this was not an issue with your definition but with ours. Basically injecting `T | undefined` while we actually only inject `T` if it is defined which is different when using spread syntax.